File: LCATest.py
import unittest
from LCA import Node
from LCA import lca
from LCA import treeToString


class TestLCA(unittest.TestCase):
    def test__single_node_tree(self):

        root = Node(1)
        assert lca(root, 1, 6).data == 1
       
    def test_seven_node_tree(self):
		root = Node(1)
		root.left = Node(2)
	    root.right = Node(3)
		root.left.left = Node(4)
		root.left.right = Node(5)
		root.right.left = Node(6)
	    root.right.right = Node(7)
        
        assert lca(root, 4, 5).data == 2
        assert lca(root, 1, 2).data == 1

    def test_lca_output_teen_tree(self):
        root = Node(15)
        root.left = Node(19)
        root.left.left = Node(12)
        root.left.right = Node(13)
        root.right = Node(14)
        root.right.left = Node(17)

        assert lca(root, 17, 13).data == 15

    def test_tree_to_string(self):
        root = Node(20)
        root.left = Node(8)
        root.right = Node(22)
        root.left.left = Node(4)
        root.left.right = Node(12)
        root.left.right.left = Node(10)
        root.left.right.right = Node(14)

        string = []
        treeToString(root, string)

        assert ''.join(string) == '20(8(4)(12(10)(14)))(22)'

    def test_LCA_none_node(self):
       
        assert lca(None, 1, 3) is None

    def test_arguments_not_in_tree(self):

        root = Node(2)
        root.left = Node(1)
        root.right = Node(3)

        assert lca(root, 5, 6) is None

    def test_empty_tree(self):

        root = None
        assert lca(root, 8, 3) is None

    def test_tree_out_of_order(self):
        root = Node(50)
        root.left = Node(30)
        root.right = Node(10)
        root.left.left = Node(40)
        root.left.right = Node(70)
        root.right.left = Node(20)
        root.right.right = Node(60)

        assert lca(root, 60, 40).data == 50
        assert lca(root, 60, 20).data == 50
        assert lca(root, 70, 20).data == 50
        assert lca(root, 30, 40).data == 30
       

File: LCATest.java
import static org.junit.Assert.assertEquals;
import org.junit.jupiter.api.Test;

class LCATest {

	@Test
	public void testEmptyTree() {
		LCA tree = new LCA();
		assertEquals("LCA of empty tree:", -1, tree.functionLCA(0, 0));
	}
	
	@Test
	public void testOneNodeTree() {
		LCA tree = new LCA();
		tree.root = new Node(1);
		assertEquals("LCA of a one node tree:", -1, tree.functionLCA(1, 0));
	}
	
	@Test
	public void testSevenNodeTree() {
		LCA tree = new LCA();
		tree.root = new Node(1);
		tree.root.left = new Node(2);
		tree.root.right = new Node(3);
		tree.root.left.left = new Node(4);
		tree.root.left.right = new Node(5);
		tree.root.right.left = new Node(6);
		tree.root.right.right = new Node(7);
		
		assertEquals("LCA of 4 and 5:", 2, tree.functionLCA(4, 5));
		assertEquals("LCA of 1 and 2:", 1, tree.functionLCA(1, 2));
		assertEquals("LCA of 6 and 7:", 3, tree.functionLCA(6, 7));	
		assertEquals("LCA of 2 and 4:", 2, tree.functionLCA(2, 4));
		assertEquals("LCA of 2 and 3:", 1, tree.functionLCA(2, 3));
		

	}
	
	@Test
	public void testThirteenNodeTree() {
		LCA tree = new LCA();
		tree.root = new Node(1);
		tree.root.left = new Node(2);
		tree.root.right = new Node(3);
		tree.root.left.left = new Node(4);
		tree.root.left.right = new Node(5);
		tree.root.right.left = new Node(6);
		tree.root.right.right = new Node(7);
		tree.root.left.left.left = new Node(8);
		tree.root.left.left.right = new Node(9);
		tree.root.left.right.left = new Node(10);
		tree.root.left.right.right = new Node(11);
		tree.root.right.left.left = new Node(12);
		tree.root.right.left.right = new Node(13);
		
		
		assertEquals("LCA of 4 and 13: ", 1, tree.functionLCA(4, 13));
		assertEquals("LCA of 8 and 9: ", 4, tree.functionLCA(8, 9));
		assertEquals("LCA of 2 and 3: ", 1, tree.functionLCA(2, 3));
		assertEquals("LCA of 1 and 12: ", 1, tree.functionLCA(1, 12));
		assertEquals("LCA of 12 and 13: ", 6, tree.functionLCA(12, 13));
		
	}
	
	@Test
	public void testNonExistentNode() {
		LCA tree = new LCA();
		tree.root = new Node(1);
		tree.root.left = new Node(2);
		tree.root.right = new Node(3);
		tree.root.left.left = new Node(4);
		tree.root.left.right = new Node(5);
		tree.root.right.left = new Node(6);
		tree.root.right.right = new Node(7);
		
		
		assertEquals("LCA of 2 and 904:", -1, tree.functionLCA(2, 904));
		assertEquals("LCA of 4 and 38:", -1, tree.functionLCA(4, 38));
	}
	
	@Test
	public void testOutOfOrder() {
		LCA tree = new LCA();
		tree.root = new Node(50);
		tree.root.left = new Node(30);
		tree.root.right = new Node(10);
		tree.root.left.left = new Node(40);
		tree.root.left.right = new Node(70);
		tree.root.right.left = new Node(20);
		tree.root.right.right = new Node(60);
		
		assertEquals("LCA of tree is out of order", 50, tree.functionLCA(60, 40));
		assertEquals("LCA of tree is out of order", 50, tree.functionLCA(70, 20));
		assertEquals("LCA of tree is out of order", 30, tree.functionLCA(30, 40));
		assertEquals("LCA of tree is out of order", 10, tree.functionLCA(60, 20));
	}
	
	@Test
	public void testSameNodes() {
		LCA tree = new LCA();
		tree.root = new Node(8);
		tree.root.left = new Node(8);
		tree.root.right = new Node(8);
		tree.root.left.left = new Node(8);
		tree.root.left.right = new Node(8);
		
		assertEquals("LCA of 8 and 8: ", 8, tree.functionLCA(8, 8));
	}
	
	@Test 
	public void unevenTree() { 
		LCA tree = new LCA();
		tree.root = new Node(1);
		tree.root.right = new Node(2);
		tree.root.right.right = new Node(3);
		
		assertEquals("LCA of uneven tree:", 1, tree.functionLCA(3, 1));
		assertEquals("LCA of uneven tree:", 2, tree.functionLCA(2, 3));	
		
	}
}
